storage/badger/operation/receipts.go

```diff
@@ -18,39 +18,42 @@ func RetrieveExecutionReceiptMeta(receiptID flow.Identifier, meta *flow.Executio
 
 // IndexExecutionReceipt inserts an execution receipt ID keyed by block ID
 func IndexExecutionReceipt(blockID flow.Identifier, receiptID flow.Identifier) func(*badger.Txn) error {
-	return insert(makePrefix(codeBlockExecutionReceipt, blockID), receiptID)
+	return insert(makePrefix(codeOwnBlockReceipt, blockID), receiptID)
 }
 
 // LookupExecutionReceipt finds execution receipt ID by block
 func LookupExecutionReceipt(blockID flow.Identifier, receiptID *flow.Identifier) func(*badger.Txn) error {
-	return retrieve(makePrefix(codeBlockExecutionReceipt, blockID), receiptID)
+	return retrieve(makePrefix(codeOwnBlockReceipt, blockID), receiptID)
 }
 
-// IndexExecutionReceipt inserts an execution receipt ID keyed by block ID and execution ID
-func IndexExecutionReceiptByBlockIDExecutionID(blockID, executorID, receiptID flow.Identifier) func(*badger.Txn) error {
-	return insert(makePrefix(codeExecutionIDExecutionReceipt, blockID, executorID), receiptID)
+// IndexExecutionReceipts inserts an execution receipt ID keyed by block ID
+func IndexExecutionReceipts(blockID, receiptID flow.Identifier) func(*badger.Txn) error {
+	return insert(makePrefix(codeAllBlockReceipt, blockID, receiptID), struct{}{})
 }
 
-// LookupExecutionReceipt finds execution receipt ID by block ID and execution ID
-func LookupExecutionReceiptByBlockIDExecutionID(blockID flow.Identifier, executorID flow.Identifier, receiptID *flow.Identifier) func(*badger.Txn) error {
-	return retrieve(makePrefix(codeExecutionIDExecutionReceipt, blockID, executorID), receiptID)
-}
-
-// LookupExecutionReceipt finds execution receipt ID by block ID for all execution IDs
-func LookupExecutionReceiptByBlockIDAllExecutionIDs(blockID flow.Identifier, receiptIDs *[]flow.Identifier) func(*badger.Txn) error {
+// LookupExecutionReceipts finds all execution receipts by block ID
+func LookupExecutionReceipts(blockID flow.Identifier, receiptIDs *[]flow.Identifier) func(*badger.Txn) error {
 	iterationFunc := receiptIterationFunc(receiptIDs)
-	return traverse(makePrefix(codeExecutionIDExecutionReceipt, blockID), iterationFunc)
+	return traverse(makePrefix(codeAllBlockReceipt, blockID), iterationFunc)
 }
 
 // receiptIterationFunc returns an in iteration function which returns all receipt IDs found during traversal
 func receiptIterationFunc(receiptIDs *[]flow.Identifier) func() (checkFunc, createFunc, handleFunc) {
 	return func() (checkFunc, createFunc, handleFunc) {
+		var val flow.Identifier
 		check := func(key []byte) bool {
-			return true
+			if len(key) == len(val)*2+1 {
+				offset := len(val) + 1
+				copy(val[:], key[offset:])
+				return true
+			} else {
+				return false
+			}
 		}
```

Comment:
can you explain how this works?

Reply:
kinda dark magic but not sure if there is another way.
We build our prefix like this:
```
func makePrefix(code byte, keys ...interface{}) []byte {
	prefix := make([]byte, 1)
	prefix[0] = code
	for _, key := range keys {
		prefix = append(prefix, b(key)...)
	}
	return prefix
}
```

it's basically concatenation of bytes. Since we are indexing by `blockID` and `receiptID` our pair is:
```
insert(makePrefix(codeAllBlockReceipt, blockID, receiptID), struct{}{})
```

When traversing badger db, I am checking for the first part of prefix: 
```
traverse(makePrefix(codeAllBlockReceipt, blockID), iterationFunc)
```
and when I have a partial match I extract the receiptID from it.
Hacky, I know, but it's the only way as I know with nosql databases.

Reply:
Thanks for the explanation @durkmurder . From my perspective, we can take two possible implementation approaches:

#### option 1 favours performance (Yurii' approach)
Yurii's approach has minimal memory footprint in the DB. If we are favouring performance, I think we should apply a few more optimizations:
* All the information we are interested in is already contained in the key, and all necessary processing steps can be done in the `check` method
   * In `check`, you already extract the relevant bytes from the key, why not append them to `receiptIDs` right away?
   * For each key, we decode the value: https://github.com/onflow/flow-go/blob/82113bba53e963f96dd235ce15cd910dcd73874c/storage/badger/operation/common.go#L293-L310 which is entirely unnecessary. If the `check` function returns false, [`traverse` would skip](https://github.com/onflow/flow-go/blob/82113bba53e963f96dd235ce15cd910dcd73874c/storage/badger/operation/common.go#L350-L353) all this unnecessary work.
* For each receipt, `traverse` calls the provided iterator function https://github.com/onflow/flow-go/blob/82113bba53e963f96dd235ce15cd910dcd73874c/storage/badger/operation/common.go#L345-L346), which re-creates `check`, `create` and `handle`. I don't think this is necessary. These functions can be created _once_ at the beginning of the query and then and reused. 
* Lastly, I think we can avoid the length check. The _current_ implementation essentially guarantees that this check is always satisfied. If the code should be inconsistently modified in the future, some tests should immediately fail, as we will be unable to retrieve previously-stored receipts. If we don't have such tests, lets implement them. This allows us to protect against future inconsistent code modifications at software creation time -- as opposed to a runtime check, which will cost us performance cost on each call.

Putting it all together, I would suggest to use the following implementation
```golang
func receiptIterationFunc(receiptIDs *[]flow.Identifier) func() (checkFunc, createFunc, handleFunc) {
	offset := len(flow.Identifier{}) + 1

	check := func(key []byte) bool {
		var val flow.Identifier
		copy(val[:], key[offset:])
		*receiptIDs = append(*receiptIDs, val)
		return false
	}

	var dummy struct{}
	create := func() interface{} { return &dummy }
	handle := func() error { return nil }

	return func() (checkFunc, createFunc, handleFunc) {
		return check, create, handle
	}
}
``` 
Arguably, this is a little bit hacky. 

#### option 2 favours readability
From my perspective, the complexity in option 1 originates from the fact that we are essentially using the key to store the desired information (not the value). Alternatively, if we want less hacky code with better readability (at the cost of some performance overhead), we could:
1. only use `codeAllBlockReceipts`, `blockID` as prefix and store a _slice_ of receipt IDs
   * the cost here is that we serialize and de-serialize _all_ the receipt IDs every time we add a new receipt ID
   * reads of all receipt IDs for a block are probably equally fast (if not even faster, as we only have to look up one key)
   * overall disk consumption would also be lower, as we still only store one IDs for each receipt , but only the block ID once.  
2. Alternatively, we can store the receipt IDs as value (which is redundant):
   ``` golang
   func IndexExecutionReceipts(blockID, receiptID flow.Identifier) func(*badger.Txn) error {
   	return insert(makePrefix(codeAllBlockReceipts, blockID, receiptID), receiptID)
   }
   ```
   But it is really readable, as it follows the established convention.

Reply:
Personally I like:
`insert(makePrefix(codeAllBlockReceipts, blockID, receiptID), receiptID)` the most. But I couldn't just waste memory so did `insert(makePrefix(codeAllBlockReceipts, blockID, receiptID), struct{}{})` since logically it makes sense in the end we are indexing by two keys. I don't really mind and go either way. So someone needs to make a decision :)

Reply:
lets waste memory. For a rate of `1 block/s`, 8 Execution nodes, we would wast about 500MB a _year_. I think that is totally acceptable.